### PR TITLE
Managing dimensionality

### PR DESCRIPTION
**Pull request by rasmusmarak**
# Description

Summary of changes

* Adapted UDP for single spacecraft optimization instead of optimizing multiple spacecraft simultaneously.
* Separated the boolean tensor from the fitness function and UDP in general. 
* Added a function to create a boolean tensor initiated with false everywhere.
* Added a function to update boolean values in the boolean tensor (used only for the champion chromosome for each udp).
* Adapted toss.py so that we optimize the trajectories for a single spacecraft at the time, where the boolean tensor is updated with each champion chromosome and then passed to the next udp.
* Updated tests that related to the fitness function for computing coverage (as we are now creating a boolean tensor outside the optimization process).
* updated notebook for plotting results (not part of module)

## Resolved Issues

- Reduced number of variables per optimization process for a swarm to equal the number of variables corresponding to optimizing one spacecraft (which hopefully lead to better swarm results)

## How Has This Been Tested?

- The code is tested with the coverage test, which has been updated format-wise with respect to the new tensor scheme.
- A new coverage test where we fill one hemisphere with true values and the other with false and then testing coverage for a single point defined in each hemisphere at the time.

### CODE REVIEW

File: plot_results.ipynb
Comment (gomezzz) [inline plot_results.ipynb:1]: what do these changes have to do with dimensionality? seem this should be another PR?

File: toss/optimization/udp_initial_condition.py
Comment (gomezzz) [inline toss/optimization/udp_initial_condition.py:1]: what do these changes have to do with dimensionality? seem this should be another PR?

File: toss/optimization/setup_parameters.py
Comment (gomezzz) [inline toss/optimization/setup_parameters.py:1]: what do these changes have to do with dimensionality? seem this should be another PR?

**Review by gomezzz — COMMENTED** [review]
Sorry but I don't understand most of the changes here and also the underlying code seems to have changed a lot without me having looked at it since #41 was merged without review. No point in reviewing like that 🤷

File: plot_results.ipynb
Comment (rasmusmarak) [reply plot_results.ipynb:1]: A comment on the merge. I apologize for the mess that it created, I should have sent a review first. 
https://github.com/rasmusmarak/TOSS/pull/41#issuecomment-1618232461

Maybe the naming scheme for this PR is wrong, I will try to think about another one. The idea was to reduce the number of dimensions for the optimization process by optimizing one spacecraft at the time, keeping the previous coverage in memory to guide the optimization of the upcoming spacecraft, as we discussed last week. The changes in the code that can be seen in the commits above are the required changes for optimizing in that way. More specifically, creating the boolean tensor outside the udp, in order to the update it with the optimal trajectory and pass the new tensor to the next udp for the optimization of the second spacecraft etc. 

The last five commits where the corresponding changes for the test functions as the input/output was a bit changed now that we're also passing a tensor etc.

**Comment by rasmusmarak** [comment]
New description of the PR can be found above.

File: toss/fitness/fitness_function_utils.py [hunk 1]
```diff
@@ -143,25 +117,140 @@ def compute_space_coverage(number_of_spacecrafts: int, spin_axis: np.ndarray, sp
...
+        i = np.argmin(np.abs(r[:, np.newaxis] - r_points), axis=0) # indices along r axis
+        j = np.argmin(np.abs(theta[:, np.newaxis] - theta_points), axis=0) # indices along theta axis
+        k = np.argmin(np.abs(phi[:, np.newaxis] - phi_points), axis=0) # indices along phi axis
+
+        # Update tensor with information on new trajectory using advanced indexing
+        bool_tensor[i, j, k] = True
+        return bool_tensor
+
+
+def get_spherical_tensor_grid(time_step: int, radius_min: float, radius_max: float, max_velocity_scaling_factor: float) -> Union[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
```
Comment (gomezzz) [inline toss/fitness/fitness_function_utils.py:207]: create_spherical_tensor_grid?
init_spherical_tensor_grid?

"get" usually means it exists already

File: toss/fitness/fitness_function_utils.py [hunk 2]
```diff
@@ -143,25 +117,140 @@ def compute_space_coverage(number_of_spacecrafts: int, spin_axis: np.ndarray, sp
...
+        radius_max (float): Outer radius of spherical grid, typically radius_outer_bounding_sphere.
+        max_velocity_scaling_factor (float): Scales the magnitude of the fixed-valued maximal velocity and therefore also the grid spacing.
+
+    Returns:
+        r (np.ndarray): Array of r coordinates for each point defined on the spherical tensor.
+        theta (np.ndarray): Array of theta coordinates for each point defined on the spherical tensor.
+        phi (np.ndarray): Array of phi coordinates for each point defined on the spherical tensor.
+        bool_tensor (np.ndarray): Boolean array corresponding to each point defined on the spherical tensor.
+    """
+
```
Comment (gomezzz) [inline toss/fitness/fitness_function_utils.py:226]: some asserts on the input variables wouldn't hurt

`assert timesteps > 0` etc

File: toss/fitness/fitness_function_utils.py [hunk 3]
```diff
@@ -143,25 +117,140 @@ def compute_space_coverage(number_of_spacecrafts: int, spin_axis: np.ndarray, sp
...
+    # Fixed maximal velocity from previously defined trajectory. 
+    # We use a fixed value to avoid prioritizing higher velocities. 
+    #
+    # NOTE: The fitness value for covered volume will not be feasible if
+    #        its maximal velocity exceeds the provided value below as the grid
+    #        spacing will be too small. Please use the scaling factor to adapt for this.
+    fixed_velocity = np.array([-0.02826052, 0.1784372, -0.29885126])
```
Comment (gomezzz) [inline toss/fitness/fitness_function_utils.py:233]: This should probably be in a CFG / constants file

Comment (gomezzz) [reply toss/fitness/fitness_function_utils.py:233]: Also, could it be computed dynamically? 🤔 (the values wouldn't make sense for another body, would they?)

File: toss/fitness/fitness_function_utils.py [hunk 4]
```diff
@@ -52,17 +51,12 @@ def _compute_squared_distance(positions: np.ndarray, constant: float) -> np.ndar
     return np.sum(np.power(positions,2), axis=0) - constant**2
 
 
-def compute_space_coverage(number_of_spacecrafts: int, spin_axis: np.ndarray, spin_velocity: float, positions: np.ndarray, velocities: np.ndarray, timesteps: np.ndarray, radius_min: float, radius_max: float, max_velocity_scaling_factor: float) -> float:
+def compute_space_coverage(number_of_spacecrafts: int, spin_axis: np.ndarray, spin_velocity: float, positions: np.ndarray, velocities: np.ndarray, timesteps: np.ndarray, radius_min: float, radius_max: float, r: np.ndarray, theta: np.ndarray, phi: np.ndarray, bool_tensor: np.ndarray) -> float:
```
Comment (gomezzz) [inline toss/fitness/fitness_function_utils.py:54]: maybe would make sense to have a simple test for this? Just give it a bool_tensor where e.g. the left hemisphere is all true, and then you give it first only position in the right to see if score > 0, then only in the left to check score == 0 ?

**Review by gomezzz — CHANGES_REQUESTED** [review]
some minor suggestions

File: toss.py [hunk 5]
```diff
@@ -42,59 +44,98 @@ def load_udp(args, initial_condition, lower_bounds, upper_bounds):
...
+    
+    # Define the UDA.
+    uda.set_bfe(bfe)
+    algo = pg.algorithm(uda)
+    
+    return algo
+
 
-    # Setup BFE machinery
+def run_optimization(args, initial_conditions, lower_bounds, upper_bounds):
```
[the remark this answers is not in the document]
Comment (rasmusmarak) [reply toss.py]: Fixed! The Initial condition array (although perhaps with a bit cryptic name) was meant to be a vector corresponding to given information about the initial state. For instance, it could be a vector of initial position and/or initial velocity defined by the user. In that case, it would be recognized in the optimization process and accounted for by also removing initial position/velocity from the chromosome. I defined it such that the user could choose to either optimize the initial state or letting it be fixed to any desired values.

File: toss.py [hunk 6]
```diff
@@ -42,59 +44,98 @@ def load_udp(args, initial_condition, lower_bounds, upper_bounds):
...
+
+        # Other logs for output
+        champion_f = pop.champion_f
+        champion_x = pop.champion_x
+
+        # Update boolean tensor (using trajectory resulting from champion chromosome)
+        if len(initial_conditions[spacecraft_i]) > 0:
+            spacecraft_info = np.hstack((initial_conditions[spacecraft_i], champion_x))
+        else:
+            spacecraft_info = champion_x
```
[the remark this answers is not in the document]
Comment (rasmusmarak) [reply toss.py]: It is the champion in the case where we optimize the initial state. However, if the initial state is given as a condition by the user, it is the concatenation of the initial condition (ie state) and the chromosome of control sequences. For readability, I have now updated the name of the array.

File: toss.py [hunk 7]
```diff
@@ -42,59 +44,98 @@ def load_udp(args, initial_condition, lower_bounds, upper_bounds):
...
+
+        _, list_of_ode_objects, _ = compute_trajectory(spacecraft_info, args, compute_motion)
```
[the remark this answers is not in the document]
Comment (rasmusmarak) [reply toss.py]: Actually this one was a bit tricky. Since I couldn't figure out how to reach and manipulate the attributes of the udp class through the pygmo problem object, I had to move the updating of the tensor to outside the optimization problem. Because of this, we have to re-initialize the udp and corresponding population for each new spacecraft as we will have a new tensor to account for. Since the integration objects are managed in the udp as well, and since pygmo only returns the fitness and some statistics on the optimization process, this means that we have to re-compute the trajectory using the champion outside the optimization in order to update the tensor. Luckily we only need to do this once per spacecraft so it shouldn't take too much time (although re-initializing the population is probably not very optimal with respect to time).

File: toss.py [hunk 8]
```diff
@@ -42,59 +44,98 @@ def load_udp(args, initial_condition, lower_bounds, upper_bounds):
...
+        positions, velocities, timesteps = get_trajectory_fixed_step(args, list_of_ode_objects)
+        args.problem.bool_tensor = update_spherical_tensor_grid(args.problem.number_of_spacecrafts, args.body.spin_axis, args.body.spin_velocity, positions, velocities, timesteps, args.problem.radius_inner_bounding_sphere, args.problem.radius_outer_bounding_sphere, args.problem.tensor_grid_r, args.problem.tensor_grid_theta, args.problem.tensor_grid_phi, args.problem.bool_tensor)
+
+        # Store champion information
+        champion_f_array[spacecraft_i] = champion_f
+        champion_x_array[len(champion_x)*spacecraft_i:len(champion_x)*(spacecraft_i + 1)] = champion_x
```
[the remark this answers is not in the document]
Comment (rasmusmarak) [reply toss.py]: Fixed!

File: toss.py [hunk 9]
```diff
@@ -42,59 +44,98 @@ def load_udp(args, initial_condition, lower_bounds, upper_bounds):
...
 def main():
 
     # Setup problem parameters (as DotMaP)
     args = setup_parameters()
     
-    # Setup initial state space
-    initial_condition = [-135.13402075, -4089.53592604, 6050.17636635]*args.problem.number_of_spacecrafts # [-3645.61233166, -5634.25158292, 2883.0399209] # [-135.13402075, -4089.53592604, 6050.17636635], [-135.13402075, -4089.53592604, 6050.17636635]
-    lower_bounds, upper_bounds = setup_initial_state_domain(initial_condition, 
+    # Setup initial state (eg position)
+    initial_conditions = np.array_split([-135.13402075, -4089.53592604, 6050.17636635]*args.problem.number_of_spacecrafts, args.problem.number_of_spacecrafts) # [-3645.61233166, -5634.25158292, 2883.0399209] # [-135.13402075, -4089.53592604, 6050.17636635], [-135.13402075, -4089.53592604, 6050.17636635]
```
[the remark this answers is not in the document]
Comment (rasmusmarak) [reply toss.py]: I have now removed the comments. The numbers above represent the initial position for the optimization and was a position I found previously by evaluating a number of trajectories and choosing a position that seemed to end up in a more periodic trajectory.  (It was the results of the discussion we had previously on fixing the initial position, but I will add a comment on it in the cfg for readability).

File: toss/fitness/fitness_function_utils.py [hunk 10]
```diff
@@ -143,25 +117,140 @@ def compute_space_coverage(number_of_spacecrafts: int, spin_axis: np.ndarray, sp
...
+
+        rot_pos_arr = np.empty((pos_arr.shape))
+
+        for col in range(0,len(pos_arr[0,:])):
+            rot_pos_arr[:,col] = rotate_point(timesteps[col], pos_arr[:,col], spin_axis, spin_velocity)
+
+        if counter == 0:
+            rotated_positions = rot_pos_arr
+        else:
+            rotated_positions = np.hstack((rotated_positions, rot_pos_arr))
```
[the remark this answers is not in the document]
Comment (rasmusmarak) [reply toss/fitness/fitness_function_utils.py]: I'm not completely sure on how to vectorize it, but it would be way better if possible. I think this could perhaps be done in another PR as a new feature, especially since it might be a little more complex. The issue is that we will have to create a "rotation object" for each individual angle of rotation as the pyquaternion package can only manage multiple rotations for the same angle. However, perhaps this could be done by creating all of the quaternion objects and converting them to their corresponding rotation matrices, stack these matrices and multiply with the array of positions. But I'm not sure if it would save any significant time. What do you think?

File: toss/fitness/fitness_function_utils.py [hunk 11]
```diff
@@ -143,25 +117,140 @@ def compute_space_coverage(number_of_spacecrafts: int, spin_axis: np.ndarray, sp
...
+        return bool_tensor
+    else: 
+        # Find the indices of the closest values in the meshgrid for each point using broadcasting
+        i = np.argmin(np.abs(r[:, np.newaxis] - r_points), axis=0) # indices along r axis
+        j = np.argmin(np.abs(theta[:, np.newaxis] - theta_points), axis=0) # indices along theta axis
+        k = np.argmin(np.abs(phi[:, np.newaxis] - phi_points), axis=0) # indices along phi axis
+
+        # Update tensor with information on new trajectory using advanced indexing
+        bool_tensor[i, j, k] = True
+        return bool_tensor
```
[the remark this answers is not in the document]
Comment (rasmusmarak) [reply toss/fitness/fitness_function_utils.py]: Fixed.

File: toss/fitness/fitness_function_utils.py (hunk 1 above)
Comment (rasmusmarak) [reply toss/fitness/fitness_function_utils.py]: Fixed.

File: toss/fitness/fitness_function_utils.py (hunk 2 above)
Comment (rasmusmarak) [reply toss/fitness/fitness_function_utils.py:255]: Fixed.

File: toss/fitness/fitness_function_utils.py (hunk 3 above)
Comment (rasmusmarak) [reply toss/fitness/fitness_function_utils.py]: I have for now added it to the cfg as I wanted to discuss how to compute it dynamically first. One idea for computing it dynamically could be to solve a smaller optimization problem (eg less generations etc) minimizing the squared distance to the safety radius (where the velocity is maximal in the region of interest) to get som feasible trajectory around the considered body, and then sample a velocity as close as possible to the safety radius to be used as the fixed_velocity considered above. This would be fairly easy to implement as we only need to pass the squared_distance fitness function to the udp instead of coverage. However, since this is another feature in its own, maybe it should be added in another PR?

File: toss/fitness/fitness_function_utils.py (hunk 4 above)
Comment (rasmusmarak) [reply toss/fitness/fitness_function_utils.py:54]: I have now added the test (and made some updates in the computation of unique visits).

**Review by gomezzz — APPROVED** [review]
:ok_hand:
